runtime/sema/check_transaction_declaration.go

```diff
@@ -83,13 +83,30 @@ func (checker *Checker) VisitTransactionDeclaration(declaration *ast.Transaction
 		declaration.PostConditions,
 		VoidType,
 		func() {
-			checker.withSelfResourceInvalidationAllowed(func() {
-				checker.visitTransactionExecuteFunction(declaration.Execute, transactionType)
-			})
+			checker.withResourceFieldInvalidationAllowed(
+				func(expression *ast.MemberExpression) *Member {
+					// In addition to the behaviour in composite destructors,
+					// where a self member may be invalidated,
+					// also allow invalidation of role members
+					memberInfo, _ := checker.Elaboration.MemberExpressionMemberInfo(expression)
+					if _, ok := memberInfo.AccessedType.(*TransactionRoleType); ok {
+						return memberInfo.Member
+					}
+
+					return checker.accessedSelfMember(expression)
```

Comment:
The default, like in https://github.com/onflow/cadence/pull/2262/files#diff-c34bee2125d8f3673160cc7b3c912540ad811b4622a66c3a2ed6c25fc94e1cb0R176